architecture_records/004-dataloader-v2.md

```diff
@@ -0,0 +1,314 @@
+a module but not via the implementation of its `main` functionality.
+
+Please note that our implementation needs to support certain predefined built-in handlers like `apply_chat_template`
+or `tokenize` which user can request just by a name.
+
+For example see this implementation - https://github.ibm.com/ai4code-wisdom/platform/blob/main/modelops/modelops/train.py#L251
+
+# Implementation of HuggingFace Data Loader.
+
+HF Dataloader or a dataloader which used Hugging Face API is implemented as an instance of the `Dataloader` class.
+
+When the dataloader goes through each `DataSetConfig`
+
+The HF dataloader implements functionality to process different type of files via its `load_dataset` factory.
+If not supported automatically via this, we can look to extend the factory to use an other type of interest via
+`Dataset.from_generator(<generator>)` functionality.
+
+The reason to make use of HF factory is to keep our code simple and free of duplication with preexisting library code.
+
+This also means that any implementation like `get_json_object` which load `json(l)` and then return a custom json dict
+can be implemented as data handlers over basic json dataloader from HF.
+
+### Splitting and Interleaving datasets
+
+Other argument such as `splitter_arguments` can be passed to HF [`datasets.test_train_split`](https://huggingface.co/docs/datasets/v3.0.1/en/package_reference/main_classes#datasets.Dataset.train_test_split) to create a test/train split of
+the dataset as requested by the user.
+
+In case of multiple datasets the user can request how the datasets are to be interleaved.
+The probabilies specified by users in the config `sampling.ratio` can be collected from individual datasets and passed to
+[`datasets.interleave_datasets`](https://huggingface.co/docs/datasets/v3.0.1/en/package_reference/main_classes#datasets.interleave_datasets).
+
+### Streaming datasets
+
+In case of Hugging Face Dataloader the `streaming` argument can be handled by using `IterableDatasets` instead of standard `Datasets`.
+HF provides same APIs like `datasets.interleave_datasets` over the `Iterable` datasets as well.
+
+Further important thing to note is in case of HF, if we use hugging face the `map` functionality which we use to implement data handling
+is handled in a lazy fashion meaning we don't need to handle the data handlers in a different way for streaming data. [More Information on HF Page.](https://huggingface.co/docs/datasets/en/about_mapstyle_vs_iterable#eager-data-processing-and-lazy-data-processing)
+
+## Handling data collators.
+
+Data collators specifically for TRL use cases like chat based interactions which apply chat templates and proper attention masking on the
+tokenized data like in the case of `DataCollatorForCompletionOnlyLM` handle a specific functionality on the data. In this design we consider two approaces for data collators.
+
+1. Implementing data collators as map.
```

Comment:
I feel one simplification, is that if the dataset is preprocessed, then just use a datacollator that collates all the tokens like a `DataCollatorForSeq2Seq`. 

Once data is processed, you no longer need a fancy data collator anymore. All it needs to do is form the batch